src/block/block_read.c

```diff
@@ -199,49 +200,89 @@ __wt_block_read_off(WT_SESSION_IMPL *session, WT_BLOCK *block, WT_ITEM *buf, uin
     buf->size = size;
 
     /*
-     * Check if the chunk cache has the needed data. If there is a miss in the chunk cache, it will
-     * read and cache the data. If the chunk cache has exceeded its configured capacity and is
-     * unable to evict chunks quickly enough, it will return the error code indicating that it is
-     * out of space We do not propagate this error up to our caller; we read the needed data
-     * ourselves instead.
+     * If chunk cache is configured we want to account for the race condition where the chunk cache
+     * could have stale content, and therefore a mismatched checksum. We can also have corrupted
+     * data in the chunk cache. For those scenarios, we do not want to fail immediately, so we will
+     * retry one time.
      */
-    if (F_ISSET(&S2C(session)->chunkcache, WT_CHUNKCACHE_CONFIGURED))
-        WT_RET_ERROR_OK(
-          __wt_chunkcache_get(session, block, objectid, offset, size, buf->mem, &chunkcache_hit),
-          ENOSPC);
-    if (!chunkcache_hit)
-        WT_RET(__wt_read(session, block->fh, offset, size, buf->mem));
-
-    /*
-     * We incrementally read through the structure before doing a checksum, do little- to big-endian
-     * handling early on, and then select from the original or swapped structure as needed.
-     */
-    blk = WT_BLOCK_HEADER_REF(buf->mem);
-    __wt_block_header_byteswap_copy(blk, &swap);
-    check_size = F_ISSET(&swap, WT_BLOCK_DATA_CKSUM) ? size : WT_BLOCK_COMPRESS_SKIP;
-    if (swap.checksum == checksum) {
-        blk->checksum = 0;
-        if (__wt_checksum_match(buf->mem, check_size, checksum)) {
-            /*
-             * Swap the page-header as needed; this doesn't belong here, but it's the best place to
-             * catch all callers.
-             */
-            __wt_page_header_byteswap(buf->mem);
-            return (0);
+    retries = 1;
+    do {
```

Comment:
Thanks for re-organizing this @ruby-oujo. I hadn't thought of using a loop here, but it makes good sense!

I started with a few tactical comments. But on thinking about this I wonder if it would be simpler to read if it was structured as regular `for` or `while` loop instead of a `do ... while`. I.e., putting the loop condition at the beginning is easier for the reader.

In particular, I think it will be easier to read if one pass through the loop follows the logic of "read some data then validate the data, report any errors." So I would:

- Move the initial error case (we got bad data from the chunk cache) to the end of the loop. 
- Move the fatal error cases out of the loop, so we can simplify the flow control and avoid multiple `continue` statements. 
- Add a flag saying what kind of error we hit, so we can print the right error message if we fall out of the loop without successfully reading the data.

In pseudo-code I think it would wind up looking something like this:

```
bool full_cksum_mismatch;
int failures, max_failures;

failures = 0;
max_failures = CHUNKCACHE_CONFIGURED ? 2 : 1;

while (failures < max_failures) {
    full_cksum_mismatch = false;

    if (failures == 0)
        __wt_chunkcache_get();
    if (failures > 0 || !chunkcache_hit)
        __wt_read();

    Block header stuff

    if (swap.checksum == checksum) {
        if (__wt_checksum_match()) {
             page_header_byteswap
             return(0);  /* SUCCESS! */
        }
        full_cksum_mismatch = true;
    } 
    failures++;
    
    /* This is repeats the loop condition, but it is simpler to do it here */
    if (failures < max_failures) 
        __wt_verbose ("reloading data" message);
}

/* Use full_cksum_mismatch to decide which error message we use */
__wt_errx();

__bm_corrupt_dump()